### PR TITLE
Make readThis/writeThis/readWriteThis throw

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR makes the compiler generated and default implementation of `readThis`, `writeThis` and `readWriteThis` throw. It also updates all the overloads of these methods in the standard library to be marked as `throws`.

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none`
- [x] `ALL` on `linux64` when `CHPL_COMM=gasnet`

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
@mppf

So I'm on the last part of this task, which is triggering compiler errors when one class in an inheritance hierarchy throws but not all of them do.

Where do you think the best place to put this code would be? During resolution of individual methods or at the end? Should I add this check only for `writeThis`?

**Comment by mppf** [comment]
> So I'm on the last part of this task, which is triggering compiler errors when one class in an inheritance hierarchy throws but not all of them do.

I was expecting that the compiler would already do this. It should be checking it in virtualDispatch.cpp somewhere.

**Comment by dlongnecke-cray** [comment]
@mppf

It looks like there is still some work left to do that I did not account for. The following compiles and yields output that could be confusing. 

```chapel

class A {
  var x: int = 0;
}

class A1: A {
  var y: int = 0;

  proc writeThis(ch: channel) {
    ch <~> "GOODBYE";
  }
}

proc main() {
  var a: A = new A();
  var b: A1 = new A1();
  var c: A = new A1();

  writeln(a);
  writeln(b);
  writeln(c);
}
```
```
{x = 0}
GOODBYE
{x = 0}
```

Should we expect this to issue an override error instead?

**Comment by dlongnecke-cray** [comment]
For now, I could merge the current work and document that failure in a future.

**Comment by mppf** [comment]
> Should we expect this to issue an override error instead?

I think the program you showed should do that but I think override checking is probably currently off for `writeThis`. The output it produced looks right to me if the `override` keyword is added. Making a future is appropriate.

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -3090,7 +3090,13 @@ private inline proc _read_one_internal(_channel_internal:qio_channel_ptr_t,
   // to stop reading if there was an error.
   qio_channel_clear_error(_channel_internal);
 
-  x.readThis(reader);
+  try {
+    x.readThis(reader);
+  } catch err {
+    //
+    // TODO: What to do with the caught error? Propagate back up?
+    //
```
Comment (mppf) [inline modules/standard/IO.chpl]: Please set the error on the channel using qio_channel_set_error etc to *something* for now

File: modules/standard/IO.chpl [hunk 2]
```diff
@@ -3120,26 +3126,32 @@ private inline proc _write_one_internal(_channel_internal:qio_channel_ptr_t,
...
     } else {
-      // ddata / cptr
       x.writeThis(writer);
     }
-  } else {
-    x.writeThis(writer);
+  } catch err {
+    //
+    // TODO: What to do with the caught error? Propagate back up?
+    //
```
Comment (mppf) [inline modules/standard/IO.chpl]: same here

File: modules/standard/IO.chpl (hunk 1 above)
Comment (dlongnecke-cray) [reply modules/standard/IO.chpl]: Just to double check, this is using `_qio_channel_set_error_unlocked`?

Comment (dlongnecke-cray) [reply modules/standard/IO.chpl]: Ah, has to be, comment mentions we already have the lock (I can't read).

**Pull request merged**